staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/timing_ratio_histogram.go

```diff
@@ -0,0 +1,287 @@
+// only for the tiny period of time until NewForLabelValuesSafe sets
+// the ratio based on the given initial numerator and denominator.
+// Thus there is a tiny splinter of time during member construction when
+// its underlying TimingHistogram is given the initial numerator rather than
+// the initial ratio (which is obviously a non-issue when both are zero).
+// Note the difficulties associated with extracting a member
+// before registering the vector.
+func NewTimingRatioHistogramVec(opts *compbasemetrics.TimingHistogramOpts, labelNames ...string) *TimingRatioHistogramVec {
+	return NewTestableTimingRatioHistogramVec(time.Now, opts, labelNames...)
+}
+
+// NewTestableTimingHistogramVec adds injection of the clock.
+func NewTestableTimingRatioHistogramVec(nowFunc func() time.Time, opts *compbasemetrics.TimingHistogramOpts, labelNames ...string) *TimingRatioHistogramVec {
+	delegate := compbasemetrics.NewTestableTimingHistogramVec(nowFunc, opts, labelNames)
+	return &TimingRatioHistogramVec{
+		Registerable: delegate,
+		delegate:     delegate,
+	}
+}
+
+func (v *TimingRatioHistogramVec) metrics() Registerables {
+	return Registerables{v}
+}
+
+// NewForLabelValuesChecked will return an error if this vec is not hidden and not yet registered
+// or there is a syntactic problem with the labelValues.
+func (v *TimingRatioHistogramVec) NewForLabelValuesChecked(initialNumerator, initialDenominator float64, labelValues []string) (RatioedGauge, error) {
+	underMember, err := v.delegate.WithLabelValuesChecked(labelValues...)
+	if err != nil {
+		return noopRatioed{}, err
+	}
+	underMember.Set(initialNumerator / initialDenominator)
+	return &timingRatioHistogramInner{
+			delegate:    underMember,
+			numerator:   initialNumerator,
+			denominator: initialDenominator,
+		},
+		nil
+}
+
+// NewForLabelValuesSafe is the same as NewForLabelValuesChecked in cases where that does not
+// return an error.  When the unsafe version returns an error due to the vector not being
+// registered yet, the safe version returns an object that implements its methods
+// by first calling WithLabelValuesChecked and then delegating to whatever observer that returns.
+// In the other error cases the object returned here is a noop.
+func (v *TimingRatioHistogramVec) NewForLabelValuesSafe(initialNumerator, initialDenominator float64, labelValues []string) RatioedGauge {
+	tro, err := v.NewForLabelValuesChecked(initialNumerator, initialDenominator, labelValues)
+	if err == nil {
+		return tro
+	}
+	if !compbasemetrics.ErrIsNotRegistered(err) {
+		klog.ErrorS(err, "Failed to extract TimingRatioHistogramVec member, using noop instead", "vectorname", v.FQName(), "labelValues", labelValues)
+		return tro
+	}
+	return &timingRatioHistogramInner{
+		delegate:    &timingHistogramVecElt{delegate: v.delegate, labelValues: labelValues},
+		numerator:   initialNumerator,
+		denominator: initialDenominator,
+	}
+}
+
+type timingHistogramVecElt struct {
```

Comment:
What is timingHistogramVecElt?

Reply:
I added a comment explaining it.  It is in service of NewForLabelValuesSafe, whose comment I noticed was buggy, so I fixed that too.

Reply:
Revised to spell it out.